extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java

```diff
@@ -1065,62 +1000,149 @@ private JsonObject getVideoSecondaryInfoRenderer() throws ParsingException {
+                    builder.setDeliveryMethod(DeliveryMethod.DASH);
+                }
 
-        // Use the mobileStreamingData object first because there is no n param and no
-        // signatureCiphers in streaming URLs of the Android client
-        urlAndItags.putAll(getStreamsFromStreamingDataKey(
-                mobileStreamingData, streamingDataKey, itagTypeWanted));
-        urlAndItags.putAll(getStreamsFromStreamingDataKey(
-                desktopStreamingData, streamingDataKey, itagTypeWanted));
+                return builder.build();
+            }
+        };
+    }
+    @Nonnull
+    private StreamBuilderHelper<VideoStream> getVideoStreamBuilderHelper(
+            final boolean areStreamsVideoOnly) {
+        return new StreamBuilderHelper<VideoStream>() {
+            @Nonnull
+            @Override
+            public VideoStream buildStream(@Nonnull final ItagInfo itagInfo) {
+                final ItagItem itagItem = itagInfo.getItagItem();
+                final VideoStream.Builder builder = new VideoStream.Builder()
+                        .setId(String.valueOf(itagItem.id))
+                        .setContent(itagInfo.getContent(), itagInfo.getIsUrl())
+                        .setMediaFormat(itagItem.getMediaFormat())
+                        .setIsVideoOnly(areStreamsVideoOnly)
+                        .setItagItem(itagItem);
+
+                final int height = itagItem.getHeight();
+                if (height > 0) {
+                    final StringBuilder stringBuilder = new StringBuilder();
+                    stringBuilder.append(height);
+                    stringBuilder.append("p");
+                    final int fps = itagItem.getFps();
+                    if (fps > 30) {
+                        stringBuilder.append(fps);
+                    }
+                    builder.setResolution(stringBuilder.toString());
+                } else {
+                    final String resolutionString = itagItem.getResolutionString();
+                    builder.setResolution(resolutionString != null ? resolutionString : "");
+                }
 
-        return urlAndItags;
+                if (streamType != StreamType.VIDEO_STREAM || !itagInfo.getIsUrl()) {
+                    // YouTube uses the DASH delivery method for videos on OTF streams and
+                    // for all streams of post-live streams and live streams
+                    builder.setDeliveryMethod(DeliveryMethod.DASH);
+                }
+
+                return builder.build();
+            }
+        };
     }
 
     @Nonnull
-    private Map<String, ItagItem> getStreamsFromStreamingDataKey(
+    private List<ItagInfo> getStreamsFromStreamingDataKey(
             final JsonObject streamingData,
             final String streamingDataKey,
-            final ItagItem.ItagType itagTypeWanted) {
+            final ItagItem.ItagType itagTypeWanted,
+            final StreamType streamType) throws ParsingException {
```

Comment:
I think the `streamType` param is always equal to the class attribute `streamType`. Please check if this is true and remove the param if necessary.
The method grew quite a bit. What do you think about moving everything within
``` Java
if (itagType == itagTypeWanted) {
   <THIS CODE>
}
```
into an own method? This would separate the actual data extraction and the logic whether an extraction is necessary [and solve some indentation problems :)].

Reply:
> I think the `streamType` param is always equal to the class attribute `streamType`. Please check if this is true and remove the param if necessary.

Indeed, parameter removed from the method.

> What do you think about moving everything within
> ```java
> if (itagType == itagTypeWanted) {
>   <THIS CODE>
>}
>```
> into an own method?

Done, everything in this block was moved into a new private method named `buildAndAddItagInfoToList`!